pkgs/by-name/pr/protege/package.nix

```diff
@@ -0,0 +1,110 @@
+{ stdenvNoCC
+, lib
+, fetchFromGitHub
+, copyDesktopItems
+, iconConvTools
+, makeDesktopItem
+, makeWrapper
+, jdk11
+, maven
+}:
+
+let
+  pname = "protege";
+  version = "5.6.4";
+  mvn = maven.override { jdk = jdk11; };
+
+  # Build a platform-independent core from which we can derive platform-specific
+  # directory layouts, so that optimized binary caches can de-duplicate it
+  # effectively.
+  protege-build = mvn.buildMavenPackage {
+    pname = "protege-build";
+    inherit version;
+
+    src = fetchFromGitHub {
+      owner = "protegeproject";
+      repo = "protege";
+      rev = version;
+      hash = "sha256-Q3MHa7nCeF31n7JPltcemFBc/sJwGA9Ev0ymjQhY/U0=";
+    };
+
+    patches = [
+      # Pin built-in Maven plugins to avoid checksum variations on Maven updates
+      ./enforce-plugin-versions.patch
+      # Avoid building platform-dependent builds which embed their own JREs
+      ./platform-independent-only.patch
+    ];
+
+    mvnHash = "sha256-kemP2gDv1CYuaoK0fwzBxdLTusarPasf2jCDQj/HPYE=";
+
+    installPhase = ''
+      runHook preInstall
+
+      mv protege-desktop/target/protege-${version}-platform-independent/Protege-${version} $out
+
+      runHook postInstall
+    '';
+  };
+in
+stdenvNoCC.mkDerivation rec {
+  inherit pname version;
+
+  buildInputs = [ protege-build ];
```

Comment:
This is a bit weird, perhaps it'd be possible to combine the `installPhase` of this derivation to the `installPhase` of this one?

Reply:
It is, but I wanted to keep them separate according to the following reasoning:
- the result of the Maven build contains a target-agnostic directory containing all the necessary jars plus a launcher script for every platform
- what distinguishes an installation for a certain platform from one for another platform is just how the produced files are placed in the filesystem (e.g. where the icon files are copied)

So why not produce a perfectly agnostic build that is the same for all platforms, and then produce a platform-specific symlink structure that refers to its relevant components as part of a platform-specific derivation? This way, when Darwin systems will be supported by this package, both Linux and Darwin builds would share the same core derivation and access it via symlinks, organized according to their platform-specific filesystem conventions. This can save space and, potentially, computation, as a single multi-target derivation would need to be evaluated for every end platform and would duplicate perfectly agnostic Java bytecode for no reason.

I don't know if my explanation is clear...

Reply:
> It is, but I wanted to keep them separate according to the following reasoning:
> 
>     * the result of the Maven build contains a target-agnostic directory containing all the necessary jars plus a launcher script for every platform

How are you so sure that this `protege-build` is the same for all platforms?

> 
>     * what distinguishes an installation for a certain platform from one for another platform is just how the produced files are placed in the filesystem (e.g. where the icon files are copied)

I'm not sure you gain anything with this. Do you want to save a few kilobytes on Hydra? What's important, is that both platforms have the same `mvnHash`.

> So why not produce a perfectly agnostic build that is the same for all platforms, and then produce a platform-specific symlink structure that refers to its relevant components as part of a platform-specific derivation? This way, when Darwin systems will be supported by this package, both Linux and Darwin builds would share the same core derivation and access it via symlinks, organized according to their platform-specific filesystem conventions. This can save space and, potentially, computation,

Yes I agree with the computation argument, but I'm not sure it is worth the readabilty penalty, and the harder overriding capabilities.

> as a single multi-target derivation would need to be evaluated for every end platform and would duplicate perfectly agnostic Java bytecode for no reason.
> 
> I don't know if my explanation is clear...

You were clear, it's just that since the `java` compiler is different between the platforms, I suspect that the derivations won't get the same hash (which is based on their inputs).

Reply:
> P. S. It would be very interesting to investigate the sources of impurity in Java bytecode, so that we can aim at reproducible Java programs in the future. I don't know if anyone has ever made some research in that direction, but I find it likely.

That's something to mention at https://github.com/NixOS/nixpkgs/issues/313216 .

> Thank you!

You are welcome!

> On the bright side, `optimise`d stores can properly detect duplication even with differing directory structures, so all JAR dependencies are correctly hard-linked.

What is this `optimise` tool you are talking about?

Reply:
`nix-store --optimise`